src/macosx/hidjoy.m

```diff
@@ -493,11 +492,19 @@ static void value_callback(
    }
 
    int int_value = IOHIDValueGetIntegerValue(value);
+   int min = joy->min[joy->dpad_stick][1];
+   int max = joy->max[joy->dpad_stick][1];
 
    if (joy->dpad == elem){
-      if (int_value >= 0 && int_value < MAX_HAT_DIRECTIONS) {
-         osx_joy_generate_axis_event(joy, joy->dpad_stick, joy->dpad_axis_vert,  (float)hat_mapping[int_value].axisV);
-         osx_joy_generate_axis_event(joy, joy->dpad_stick, joy->dpad_axis_horiz, (float)hat_mapping[int_value].axisH);
+      if (int_value >= min && int_value <= max) {
+         int index = int_value - min;
+         if (index < MAX_HAT_DIRECTIONS) {
+            osx_joy_generate_axis_event(joy, joy->dpad_stick, joy->dpad_axis_vert,  (float)hat_mapping[index].axisV);
+            osx_joy_generate_axis_event(joy, joy->dpad_stick, joy->dpad_axis_horiz, (float)hat_mapping[index].axisH);
```

Comment:
if the device has just 4 distinct states instead of 8, prob can support by just multiplying `index` by 2. But I don't have such a device so I can't test that theory.